### PR TITLE
feat: allow DERP headers to be set

### PR DESCRIPTION
**Pull request by JoshVee**
This PR adds the ability to set custom HTTP headers to be passed to the Tailscale connection. This allows the websocket-based DERP connection to function behind reverse proxies or tunnels (for example, authentication headers).

The DERP header functionality is controlled via the existing `--header` CLI flags.

This functionality utilizes the changes introduced in the following PRs:
- https://github.com/coder/tailscale/pull/6
- https://github.com/coder/tailscale/pull/12

### CODE REVIEW

**Comment by JoshVee** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by JoshVee** [comment]
recheck

**Comment by kylecarbs** [comment]
@JoshVee can we make it always use the global HTTP headers instead? I'd rather not add a new top-level flag specifically for DERP.

**Comment by JoshVee** [comment]
Thanks @kylecarbs. I've made the relevant changes.

**Comment by JoshVee** [comment]
recheck

**Comment by JoshVee** [comment]
Hi @kylecarbs,
Any chance we can get this included in the next release? This is a key feature for our use case.
Thanks.

**Comment by kylecarbs** [comment]
@JoshVee certainly!

**Comment by bpmct** [comment]
> This is a key feature for our use case.

Can you explain why you need this, is it for a specific LoadBalancer or proxy?

**Comment by kylecarbs** [comment]
@JoshVee doing a release today that will include this. I'm going to make a few minor changes on your branch if that's alright!

**Review by kylecarbs — APPROVED** [review]

File: codersdk/workspaceagents.go [hunk 1]
```diff
@@ -143,9 +143,17 @@ func (c *Client) DialWorkspaceAgent(ctx context.Context, agentID uuid.UUID, opti
 	}
 
 	ip := tailnet.IP()
+	var header http.Header
+	headerTransport, ok := c.HTTPClient.Transport.(interface {
+		Header() http.Header
+	})
+	if ok {
+		header = headerTransport.Header().Clone()
```
Comment (mafredri) [inline codersdk/workspaceagents.go:151]: We clone in both the function and here, not a big issue but thought I'd point it out.

**Review by mafredri — APPROVED** [review]
Looks like scaletest needs some changes, but other than that, LGTM! 👍🏻

File: cli/root.go [hunk 2]
```diff
@@ -655,12 +655,18 @@ func checkWarnings(cmd *cobra.Command, client *codersdk.Client) error {
...
+
+func (h *headerTransport) Header() http.Header {
+	return h.header.Clone()
 }
 
 func (h *headerTransport) RoundTrip(req *http.Request) (*http.Response, error) {
-	for k, v := range h.headers {
-		req.Header.Add(k, v)
+	for k, v := range h.header {
+		for _, vv := range v {
```
[the remark this answers is not in the document]
Comment (kylecarbs) [reply cli/root.go:667]: We probably _should_, but feels like it's not super hard to add in the future anyhow.

File: codersdk/workspaceagents.go (hunk 1 above)
Comment (mafredri) [inline codersdk/workspaceagents.go:151]: (Accidentally deleted this comment): We seem to clone in both the function and here, not an issue per-se just pointing it out.

Comment (kylecarbs) [reply codersdk/workspaceagents.go:151]: Ahh, I'll change that!

Comment (kylecarbs) [reply codersdk/workspaceagents.go]: Fixed!

**Pull request merged**